File: api/ytdl.js
const { exec } = require('child_process');
const API_KEY = 'AlphaCoder03';

module.exports = async (req, res) => {
    const { url, apikey } = req.query;
    if (apikey !== API_KEY) {
        return res.end(
            JSON.stringify(
                {
                    status: 'error',
                    errorCode: 'INVALID_API_KEY',
                    message: 'API Key tidak valid.',
                    timestamp: new Date().toISOString(),
                    requestId: apikey || 'N/A',
                    details: 'Periksa API Key yang Anda kirimkan dan pastikan itu benar.',
                },
                null,
                2
            )
        );
    }

    if (!url) {
        return res.end(
            JSON.stringify(
                {
                    status: 'error',
                    message: 'URL video tidak diberikan.',
                },
                null,
                2
            )
        );
    }

    const command = `yt-dlp -f best --get-title --get-duration --get-url --get-filename "${url}"`;

    exec(command, (error, stdout, stderr) => {
        if (error) {
            return res.end(
                JSON.stringify(
                    {
                        status: 'error',
                        message: 'Terjadi kesalahan saat menjalankan yt-dlp.',
                        details: error.message,
                    },
                    null,
                    2
                )
            );
        }

        if (stderr) {
            return res.end(
                JSON.stringify(
                    {
                        status: 'error',
                        message: 'Terjadi kesalahan pada yt-dlp.',
                        details: stderr,
                    },
                    null,
                    2
                )
            );
        }

        try {
            const output = stdout.split('\n').filter(Boolean);
            const title = output[0];
            const duration = output[1];
            const mp4Url = output[2];
            const mp3Url = output[3];

            return res.end(
                JSON.stringify(
                    {
                        status: 'success',
                        project: 'AlphaCoder',
                        owner: 'Anton',
                        data: {
                            title: title,
                            duration: duration,
                            mp4Url: mp4Url,
                            mp3Url: mp3Url,
                        },
                    },
                    null,
                    2
                )
            );
        } catch (e) {
            return res.end(
                JSON.stringify(
                    {
                        status: 'error',
                        message: 'Terjadi kesalahan dalam memproses data.',
                        details: e.message,
                    },
                    null,
                    2
                )
            );
        }
    });
};

File: api/ytdl.py
import yt_dlp
from flask import Flask, request, jsonify
app = Flask(__name__)
API_KEY = 'AlphaCoder03'

@app.route('/api/ytdl', methods=['GET'])
def ytdl():
    url = request.args.get('url')
    apikey = request.args.get('apikey')

    if apikey != API_KEY:
        return jsonify({
            "status": "error",
            "errorCode": "INVALID_API_KEY",
            "message": "API Key tidak valid.",
            "timestamp": str(request.args.get('timestamp', 'N/A')),
            "requestId": apikey or 'N/A',
            "details": "Periksa API Key yang Anda kirimkan dan pastikan itu benar."
        }), 400

    if not url:
        return jsonify({
            "status": "error",
            "message": "URL tidak diberikan."
        }), 400

    try:
        ydl_opts = {
            'quiet': True,
            'format': 'bestaudio/bestvideo',
            'outtmpl': '%(id)s.%(ext)s',
        }

        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info_dict = ydl.extract_info(url, download=False)
            title = info_dict.get('title', 'Tidak ada judul')
            duration = info_dict.get('duration', 0)
            mp4_url = None
            mp3_url = None

            for format in info_dict['formats']:
                if format['ext'] == 'mp4' and not mp4_url:
                    mp4_url = format['url']
                if format['ext'] == 'mp3' and not mp3_url:
                    mp3_url = format['url']

        return jsonify({
            "status": "success",
            "project": "AlphaCoder",
            "owner": "Anton",
            "data": {
                "title": title,
                "duration": duration,
                "mp4_url": mp4_url,
                "mp3_url": mp3_url
            }
        })

    except Exception as e:
        return jsonify({
            "status": "error",
            "message": "Terjadi kesalahan pada server.",
            "details": str(e)
        }), 500


if __name__ == "__main__":
    app.run(debug=True)